### DISCUSSION TITLE
Questions about Network analysis examples

### DISCUSSION THREAD

**Discussion by MariellaCC** [DHARPA-Project/DHARPA-Project-workflows-requirements#1] [General]
I. General

1) Would it be possible to have some kind of decision tree or process, or to know about which steps a researcher would take to decide between the different categories below? I guess that often the data will obviously impose restrictions since all methodologies can't be used with any datasets that a user may have. But when (and it does happen throughout the examples) users have data that is usable with several methodologies, what would be the thinking process/underlying questions to decide to try one methodology or the other?
Whole / Ego
One-Mode / Two-mode
Directed / Undirected
Multigraph

2) Would it be possible to list the different combinations possible of the methodologies listed above?

3) Is the distinction between Whole and Ego important (I noticed that so far it didn't come up in the first examples, but I guess it could come up in the future?) The reason why I'm mentioning this is that it is quite mentioned a lot in several reference books, for example in John Scott, Social Network Analysis - SAGE Research Methods, 2017, accessed March 11, 2022.
And I don't know if this would have implications:
"the concepts used will sometimes have a slightly different meaning or method of calculation according to whether the research is
concerned with egonets or whole networks"

II. Weighted directed example (https://github.com/DHARPA-Project/NetworkXAnalysis/blob/master/ReviewsNXPandas.ipynb)

1) Would it be possible to have all specific terms explained  (like it was done for Degree centrality, Betweenness centrality, Eigenvector centrality). For example "Diameter" or "strongly connected", how would a user interpret these metrics (how would they explain the number/result in a sentence for example) and what conclusion would they draw from computing them? Some of these notions were already mentioned in meetings, but it would be helpful to centralise the info, maybe in the readme file if some concepts are useful across workflows?

2) If the graph was not weighted, would the workflow be the same, except for the “weighted” computations?
Are there, for example, computations that don’t make sense for weighted and would make sense, if not weighted (except the specific “weighted” ones) like for example density (as suggested in this other paragraph from book mentioned above: "The density measure for valued graphs, therefore, is highly sensitive to assumptions that a researcher makes about the data. A measure of density calculated in this way, however, is totally incommensurable with a measure of density for unvalued data. For this reason, it is important that a researcher does not use a measure simply because it is available in a standard program. A researcher must always be perfectly clear about the assumptions involved in any particular procedure, and must report these along with the density
measures calculated."

3) Would it be possible to elaborate this sentence present in the example notebook (concerning density metrics): "This calculation can be helpful to determine the most useful types of visualisation for the graph"

4) Some of the steps are presented as "it is possible to do.." (for example in the "Make some calculations for undirected graphs" section, would it be possible to give context about why these calculations would make sense in the context of the dataset/research question ?)

**Comment by yaslena** [comment]
Response to I.3 "Is the distinction between Whole and Ego important?"

An ego-network can be considered a special type of subgraph of a network. If you create a subgraph, you automatically change the frame of reference for the network. Beacause you are working with a selection when you deal with a subgraph, the relationships now appear in a different light, they are presented as decontextualised from the original (whole) network. It is in the nature of selections and decontextualisations that meaning can change when compared to the original context. An ego-network is no exception here. So, in that sense, caution is indeed called for. 

As for different methods of calculations specifically for ego-networks. UCINET software, in reference to which John Scotts _Social Network Analysis_ was written, offers a variety of specific ego-network calculations (1-12):
![Screenshot 2022-03-11 at 11 25 20](https://user-images.githubusercontent.com/53467834/157849421-b5a3dd7f-d44f-4da8-9354-f9603381fb21.png)
(Source: https://faculty.ucr.edu/~hanneman/nettext/C9_Ego_networks.html 

In _kiara_ we currently don't support ego-network specific calculations, although some of the same numbers could be obtained by creating an ego graph (ego_graph(G, n, radius=1, center=True, undirected=False, distance=None)) and then running the usual calculations on that subgraph. A special case is perhaps density. The author proposes two different types of density for ego networks, one that includes the edges of the ego-node and one that excludes them:
![Screenshot 2022-03-11 at 12 36 01](https://user-images.githubusercontent.com/53467834/157859682-fd5a3a09-f08f-4c2a-a35e-59d55d8ce968.png)
(Source: John Scott, Social Network Analysis - SAGE Research Methods, 2017, accessed March 11, 2022, https://methods.sagepub.com/book/social-network-analysis-4e.)

This is a valid point, since the edges of the ego-node are always included in an ego network per definition (and are thus less interesting because taken for granted), but the more interesting density-related question could be how interconnected the other nodes are among themselves. However, when comparing multiple ego networks and their density values with one another, the "inclusive" measures would be enough to get comparative results.

**Comment by yaslena** [comment]
II.2 Different workflows for weighted and unweighted graphs and other questions

a) Yes, there are special problems for weighted graphs. An overview can be found here:
Skiena S.S. (2012) Weighted Graph Algorithms. In: The Algorithm Design Manual. Springer, London. https://doi-org.proxy.bnl.lu/10.1007/978-1-84800-070-4_6 (added to our Zotero library)

b) The density problem:
The density of a graph depends on 3 things:

- number of nodes
- actual number of edges
- maximum possible number of edges

In some cases, the **maximum possible number of edges** is difficult to define and can be a matter of interpretation or case dependent. Typical examples that cause problems are graphs with self-loops and parallel edges. If the graph has self loops or parallel edges, it is up to the user to define what the **maximum possible number of edges** should be. How many self-loops do you set as maximum for each node? 1, 5 or 20? How many parallel edges do you allow for each node? The answer to that question directly affects the **maximum possible number of edges**. John Scott extends the problem to weighted graphs. In some cases it is reasonable to consider an edge weight as an edge mulitplier. That means that the edge weight X could be translated as X parallel edges. Again you have the problem of how to calculate the maximum possible number of edges. One can take the highest existing edge weight in the graph and assume that weight as the maximum possible weight for each node. This is just one of many possible solutions; there is no perfect solution that would work for all types of networks.
Another way to get information that considers the intensity of the connections in a network is to calculate **average weighted degrees**.

**Comment by yaslena** [comment]
II.1 Network Terminology

Putting those in the readme file seems like a good idea.
These lists are good starting points:
https://cvcedhlab.hypotheses.org/106

https://folgerpedia.folger.edu/Glossary_of_network_analysis_terms

**Comment by yaslena** [comment]
II.3 Density & Visualisation

Networks that are large and extremely dense can be hard to read. Strategies to visually reduce density are suggested here:
https://cambridge-intelligence.com/visualize-large-networks/

**Reply by MariellaCC** [reply]
Thanks! from what I understand of this article, it doesn't really say that density could be used as a metric to decide on a viz type (which makes sense I guess since density can not always be compared across graphs from what I understood, so that would exclude the idea of having some kind of density threshold to decide about a visualization type for example), but it rather says that if there are performance issues to display a graph, one strategy may be to reduce density by removing duplicates etc.

**Reply by makkus** [reply]
Mabye it makes sense to add a field 'potential implementation performance' issues, or something like this? This could even be further sub-devided into backend / frontend / visualisation. I can think of a few situations where sizes of data could be prohibitive, if not handled explicitely. And even if we have no solution (or don't intend to support edge-cases where those issues arise), it might be good to explicitly state that, and let users know if they try something where this is relevant. I mean: rather give a good error message, then run out of memory or display a visualisation that is basically unusable.

**Reply by makkus** [reply]
That way, we can flag those cases, and decide whether to add it to our list of things to tackle, or not, and if, in what priority.

**Reply by MariellaCC** [reply]
ok, sure, good idea! I will add such a field

**Reply by yaslena** [reply]
As far as I can tell, I'm not talking about graph types either, so I'm not sure how to place the comment. Talking about "useful types of visualisation for the graph" is talking about visualisations, not graph types.

**Reply by MariellaCC** [reply]
yes, I meant visualization types, sorry for the confusion (I will modify in the comment)

**Reply by yaslena** [reply]
I don't think that the article is primarily about performance issues. Performance is mentioned, but it is not the main focus. The artcile describes strategies to make dense and large graphs more readable for the end user. For example by "taking away anything that’s cluttering the chart unnecessarily". Visual cluttering is a problem of readability, not of performance. "Reducing clutter" and "tidying noisy charts" are all strategies aiming at making visualisations more easily understandable and insightful. I like the idea of visually grouping and combining nodes to reduce clutter and make "hairball monster" graphs more readable. I'm not sure if we would eventually offer these types of visualisations that have a zooming, node-grouping, or node-ordering effect. If it is not within our scope, then we don't need to consider this article.

**Reply by MariellaCC** [reply]
That's an interesting point, but from my perspective here (trying to figure out about inputs for the viz), the point about performance is more important and worth keeping in mind. But what I tried to clarify initially, is whether or not computing density as a metric may help in assessing a visualization type, as suggested in the notebook, but it does not seem to be the case. Concerning the UI aspect, it's also an interesting point to keep in mind, but I think this may be something that would happen after the viz is created, and would be some kind of data "augmentation" (even if it is a reduction) operation if it is supposed to be reflected in the dataset, or, if not, this would be handled in the UI only.

**Reply by yaslena** [reply]
This discussion got a bit derailed, but I would like to come back to the initial question. Can the density of a graph be an indicator to choose one type of viz over another? In my experience, it can be. It can not be a metric for viz choice, as this is not an exact science, but it can point you in a useful direction. If you have a very dense graph, it means that there are a lot of edges because almost all nodes will be connected to each other. If you choose to visualise this as a **classic network graph**, the result will likely be that you will have a lot of crossing edge lines and it will be hard to read who is connected to whom. In that case, it might be better to choose a **matrix viz**, where the edges are orderly sorted in a matrix layout, better still: a matrix with several sorting options to look for connection patterns. A **chord diagram** would be a nightmare, because this viz is especially unreadable with dense graphs. If the high graph density is due to a strong interconnectedness within certain groups in the network, then an **arc diagram** might help reveal that particularity (assuming you can group the nodes accordingly). If there are no discernable groups, then the arc diagram will be as bad a choice as the classic network graph. 

This kind of reasoning is difficult to quantify. There is no density threshold that could give definitive answers. In some cases you would just have to try out different viz and different layouts. The problem gets worse when the graph is **both dense and large**. Then the matrix layout will be useless unless you focus on a small section of the network. The article that I suggested deals with this kind of problem. Apart from filtering, it suggests a viz that enables node grouping. I would not choose this type of viz with a sparse graph, so the density of the graph would inform my choice here.

**Reply by MariellaCC** [reply]
understood! and this is interesting, but in any case, density alone does not seem to enable, based on the metric alone, to choose a viz type beforehand (without visualizing, before passing this data as an input to the viz that gets created), i.e. we couldn't, based on a certain threshold, call a viz type rather than other, from what I understand, especially if density can't be compared across graph if I understood correctly here: "A far more fundamental problem that affects all measures of density must now be considered. This is the
fact that the density depends on the size of a graph. This prevents density measures being compared across
networks of different sizes (see Niemeijer, 1973; Friedkin, 1981; Snijders, 1981). [...] This limit on the total number of lines means that larger graphs will, other things
being equal, have lower densities than small graphs." (from same book as mentioned above). In any case, thanks for the clarification, I now have all the info I need about that point at the moment.

**Reply by yaslena** [reply]
There is one observation in this comment that is incorrect and one that is irrelevant for the issue in question. I'm also assuming that  "choose a viz type beforehand" is what is actually meant because "graph type" has nothing to do with this issue. 

It is not true that density can't be known before a graph is generated. In most cases the variables for calculating density will be known in advance. (See my general comment on how to calculate grah density: https://github.com/DHARPA-Project/DHARPA-Project-workflows-requirements/discussions/1#discussioncomment-2340061). It is true that other factors might also help with this decision, but the question was not wether density is a unique marker, but whether it can be a useful indicator. As I have argued above, it can be.

It does not matter at all for the initial question whether it is generally difficult to compare density between different graphs. Even if it is problematic to compare 2 different graphs, you can still use the metric to make assumptions whether the graphs are very dense, moderately dense, a little dense, or no dense at all. Quoted out of context, the quote above might seem to be important, but it is not relevant for our discussion. A more in-depth explanation requires more in-depth knowledge of the concept of graph density. I'm happy to provide further details on that and the problem of comparing density accross graphs of different sizes (using the examples in Scott's and Niemeijer's texts), but this discussion does not belong here.

**Reply by MariellaCC** [reply]
alright, there weren't any explanations in the related notebook part, so I needed a bit of context, thanks!
how the kiara modules were designed so far and in the notebooks, I don't think density is known beforehand, but it is exactly to try to figure out this kind of things that these questions are important.

**Comment by yaslena** [comment]
II.4 Additional operations

When a step is introduced with the words "it is possible to do X..", it usually means that the step does not make sense with the sample dataset but can otherwise be insightful with the appropriate dataset. Converting the graph to undirected and then search for connected components is one such example. Knowing my data, I know that my dataset only consists of one component, so there is no incentive for me to get the computational confirmation for that fact. For other datasets, particularly large ones as in the bipartite workflow example, where there are more than 3000 connected components, it can be interesting to examine the components individually.
For some other computations, the search for connected components is a preliminary step because some algorithms only work when the graph is connected (it is a data-check, preprocessing step).

**Comment by yaslena** [comment]
I.2 Combination of graph types

- Two-mode graphs can be directed or undirected, multigraphs or not, and you can extract an ego network from them.
- A one-mode graph can be created from a bipartite two-mode graph through projection. It can be directed or undirected, a multigraph or not, and can also be further subdivided into ego-networks.
- Directed graphs can be converted into undirected graphs by ignoring the source- target meaning in the edge list.
- An ego-network can be created from any type of graph.

**Comment by yaslena** [comment]
I.1 Choosing a graph type

A related question has been asked here: https://github.com/DHARPA-Project/kiara_modules.network_analysis/discussions/18#discussioncomment-2227460

This choice is very case dependent and almost impossible to answer. Generally speaking, multi-mode graphs are perhaps better suited for exploration and one-mode graphs are better for statistical analysis. Creating a projection from a two-mode graph and obtaining a one-mode graph allows the user to apply more complex computational methods than are available for two-mode graphs. Multi-mode graphs can be used in combination with actor network theory.
Creating an ego-network allows to zoom-in on a section of the network to examine it in detail. 
Directed and undirected is rarely a choice, but when it is, the choice is usually between examining structures and examining flows. With undirected networks only the existance of connections is considered, while in directed networks the directionality is an additional attribute that can be studied to examine assymetrical connections.

- A good comparison between directed and undirected graphs: https://www.baeldung.com/cs/graphs-directed-vs-undirected-graph
- Another good source about the how and why doing network analysis in historical research is Claire Lemercier’s article: https://halshs.archives-ouvertes.fr/halshs-00521527/document
- Another guide for getting started with SNA and basic considerations: https://historicalnetworkresearch.org/first-steps/

As for deciding whether to choose a **bipartite graph**, this description from the module descussion can be useful:

> Bipartite graphs are special case of a two-mode graph. They have **two node sets** and edges that only connect nodes from opposite sets. Those types of graphs are ideal for investigating connections between two different entities, for example _authors_ and their _publications_, _politicians_ and _public debates_, _employees_ and their _work places_. This type of graph is often used in affiliation networks but can be also useful in other scenarios where the relationship between two entities is examined. If the network should contain more than two types of nodes (_authors - publications - text genres,_ or _politicians - debates - political spectrum_) then the bipartite graph type will not be suitable.

https://github.com/DHARPA-Project/kiara_modules.network_analysis/discussions/15#discussion-3870347

**Reply by MariellaCC** [reply]
Alright, thanks, as users may not be familiar with network analysis when using the tool, I think that it may be interesting and important to find a way to put this in terms of research question, and not focus too much on too technical terms in the UI: for what type of research question would one want to go for one type, or the other etc. I will add an "example research questions" field in the workflow collection template, as this is quite crucial for all workflows, and maybe by collecting several examples, a sort of pattern will emerge
